### PR TITLE
@craigspaeth - [WIP] Admin events view

### PR DESCRIPTION
**Pull request by broskoski**
screen shot 2016-10-25 at 8 36 09 pm
screen shot 2016-10-25 at 8 36 16 pm

Still haven't figured out how to wire up basic auth with unikoa :)

### CODE REVIEW

File: apps/admin/auth.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+const {
+  BASIC_AUTH_USERNAME,
+  BASIC_AUTH_PASSWORD,
+} = process.env
+
+export default function * (next) {
+  console.log('HERERERERERERE')
```
Comment (craigspaeth) [inline apps/admin/auth.js]: Stray log here

File: apps/admin/auth.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+const {
+  BASIC_AUTH_USERNAME,
+  BASIC_AUTH_PASSWORD,
+} = process.env
+
+export default function * (next) {
```
Comment (craigspaeth) [inline apps/admin/auth.js]: Woh generator function? You might just need to `npm i koa-basic-auth@next`... welcome to the fun Koa world transitioning b/t generators and async/await. Related https://github.com/koajs/basic-auth/issues/10

File: apps/admin/controllers/index.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import Transport from 'lokka-transport-http'
+import tree from 'universal-tree'
+import Index from '../views'
+import NewEvent from '../views/new_event'
+
+const eventApi = new Lokka({
+  transport: new Transport(process.env.APP_URL + '/api/event')
+})
+
+const reservationApi = new Lokka({
```
Comment (craigspaeth) [inline apps/admin/controllers/index.js]: BTW we can keep these APIs under one endpoint, when you use `graphqlize` you can pass multiple models e.g. `app.use(graphqlize(event, reservation))`

File: apps/admin/views/head.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import veact from 'veact'
+
+const reset = `
```
Comment (craigspaeth) [inline apps/admin/views/head.js]: Might make sense to extract this to lib/reset.js now

**Comment by craigspaeth** [comment]
Solid start. Auth in this stack would actually intentionally be outside of unikoa. Being that's universal routing, you wouldn't want auth concerns to be on the client-side. From what I'm noticing this could be a fine direction for an auth strategy—basically protecting an entire GraphQL API and UI under basic auth.

That said, a more diligent auth strategy would involve adding authorization (or "permissions") at the model layer to protect individual GraphQL operations, and authentication (as in login + session) at the controller/view layer. For example, if you wanted to say "creating events" was an admin-only operation you would...

1. Accept an auth token at the model layer determining the user's role is admin (let's say a JWT returned from Gravity verified/parsed with a [jwt lib](https://github.com/auth0/node-jsonwebtoken))

````javascript
import jwt from 'jsonwebtoken'
 
const event = model('event', ...)

const adminOnly = async (ctx,  next) => {
  const token = ctx.req.authToken
  jwt.verify(token, process.env.JWT_SECRET) // Throws auth error if bad token
  if (jwt.decode(token).user.type === 'Admin') next()
  else throw new Error("Only admins can create events") 
}

event.on('create', adminOnly)
````

2. OAuth with Gravity, eventually grabbing the JWT and bootstrapping into state

server.js
````javascript
import passport from 'koa-passport'

app.use(passport.initialize())
app.use(passport.session())
app.use('/login', passport.authenticate('artsy', {
  successRedirect: '/',
  failureRedirect: '/logout'
}))
app.use('/logout, controller.logout)
````

controller
````javascript
const index = (ctx, next) => {
  // ctx.user is a parsed cookie session from passport.js
  // bootstrapping it means we won't run this code on the client, and fail to find ctx.user
  const token = ctx.bootstrap(() => ctx.user.token)
  state.set('jwt', token)
  ctx.render({ ... })
}

const logout = (ctx) => ctx.logout() // Provided by passport, only used server-side
````

3. Send that token along API requests in other controller functions

````javascript
const api = new Lokka(...)

const createEvent = async (e) => {
  await api.mutate(`{
    authToken(token: ${state.get('jwt')})
    createEvent(capacity: ${e.target.value}) { capacity }
  }`)
}
````

Obviously that's missing some parts about configuring, say Koa-Passport, but I hope that illustrates the idea. The auth story could probably use some help to make it eaiser/more obvious.

**Comment by craigspaeth** [comment]
Looking good. Since you labeled WIP I'll let you decide when to merge. Happy to pair through w/e refactors/ideas sound interesting.

I hope my ideal auth story outline makes sense—given constraints I don't think we _need_ to be _that_ diligent with auth in this first run, so happy to merge and move fwd when you're ready.

File: apps/admin/views/head.js (hunk 4 above)
Comment (broskoski) [inline apps/admin/views/head.js]: Indeed

**Comment by broskoski** [comment]
Thank you @craigspaeth! Gonna clean up some stuff here, try again on basic auth (for now) and then merge-y.

**Comment by broskoski** [comment]
Mergin!

**Pull request merged**